src/main/java/Duke.java

```diff
@@ -1,10 +1,216 @@
+                fw.write(task.writeToFile());
+            }
+            fw.close();
+        } catch (IOException e) {
+            System.out.println("Something went wrong!" + e.getMessage());
+        }
+    }
+
+    // retrieve data from the file
+    public static void retrieveFromFile(String FILE_PATH) {
+        try {
+            File file = new File(FILE_PATH);
+            Scanner s = new Scanner(file);
+            Task task;
+            while (s.hasNext()) {
+                String[] parseTask = s.nextLine().split("\\|");
+                // detect what type of task it is
+                switch (parseTask[0]) {
+                case "T":
+                    task = new Todo(parseTask[2]);
+                    break;
+                case "D":
+                    task = new Deadline(parseTask[2], parseTask[3]);
+                    break;
+                case "E":
+                    task = new Event(parseTask[2], parseTask[3]);
+                default:
+                    task = new Task(parseTask[2]);
+                }
+                if(parseTask[1].equals("true")) {
+                    task.markAsDone(true);
+                }
+                tasks.add(task);
+            }
+        } catch (FileNotFoundException e) {
+            System.out.println("This file is not found!");
+        }
+    }
+
+    public static void printHorLine() {
+        System.out.println("    ____________________________________________________________");
+    }
+
+    public static void printHelloMessage() {
+        printHorLine();
+        System.out.println("     Hello! I'm Duke\n" + "     What can I do for you?");
+        printHorLine();
+    }
+
+    public static void printByeMessage() {
+        printHorLine();
+        System.out.println("     Bye. Hope to see you again soon!");
+        printHorLine();
+    }
+
+    public static void printTaskAddition(Task task, int index) {
+        printHorLine();
+        System.out.println("     Got it. I've added this task: \n" + "       " +
+                task.toString() + "\n     Now you have " +
+                index + " tasks in the list.");
+        printHorLine();
+    }
+
+    public static void printList(ArrayList<Task> tasks) {
+        printHorLine();
+        int index = 0;
+        for (Task task : tasks) {
+            System.out.println("     " + (index + 1) + "." +
+                    "     " + task.toString());
+            index++;
+        }
+        printHorLine();
+    }
+
+    public static void printCannotBeEmpty(String[] userInput) {
+        printHorLine();
+        System.out.println("     ☹ OOPS!!! The timing or description of " + Arrays.toString(userInput) + " cannot be empty.");
+        printHorLine();
+    }
+
+    public static void printInvalidMeaning() {
+        printHorLine();
+        System.out.println("     ☹ OOPS!!! I'm sorry, but I don't know what that means :-(");
+    }
+
+    // mark task as done
+    public static void markAsDone(Task task) {
```

Comment:
I have seen a structure like this across different methods: printHorLine(); print(msg); printHorLine();. Maybe you should create a method doing that, the parameter is the msg itself. With this, we avoid to repeat the same code in different sections.